### PR TITLE
Validation check for InferenceService Name

### LINKED ISSUE #1059: isvc name validation check
/kind bug
Deploy an isvc, name start with number, the kfserving manager do not check this case. This case will make the isvc won't deploy succeed, and user won't get the error msg.

* **define an isvc**
```
apiVersion: serving.kubeflow.org/v1alpha2
kind: InferenceService
metadata:
  name: 123a
spec:
  default:
    predictor:
      custom:
        container:
          name: hello
          image: iamlovingit/customized-urls:latest
```
* **deploy this isvc**
```
kubectl apply -f test.yaml 
inferenceservice.serving.kubeflow.org/123a created
```
* **check the log from kfserving manager**

```
{"level":"error","ts":1598872272.0748599,"logger":"controller-runtime.controller","msg":"Reconciler error","controller":"kfserving-controller","request":"default/123a","error":"admission webhook \"validation.webhook.serving.knative.dev\" denied the request: validation failed: not a DNS 1035 label: [a DNS-1035 label must consist of lower case alphanumeric characters or '-', start with an alphabetic character, and end with an alphanumeric character (e.g. 'my-name',  or 'abc-123', regex used for validation is '[a-z]([-a-z0-9]*[a-z0-9])?')]: metadata.name","stacktrace":"github.com/go-logr/zapr.(*zapLogger).Error\n\tgithub.com/go-logr/zapr@v0.1.1/zapr.go:128\nsigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).reconcileHandler\n\tsigs.k8s.io/controller-runtime@v0.4.0/pkg/internal/controller/controller.go:258\nsigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).processNextWorkItem\n\tsigs.k8s.io/controller-runtime@v0.4.0/pkg/internal/controller/controller.go:232\nsigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).worker\n\tsigs.k8s.io/controller-runtime@v0.4.0/pkg/internal/controller/controller.go:211\nk8s.io/apimachinery/pkg/util/wait.JitterUntil.func1\n\tk8s.io/apimachinery@v0.17.2/pkg/util/wait/wait.go:152\nk8s.io/apimachinery/pkg/util/wait.JitterUntil\n\tk8s.io/apimachinery@v0.17.2/pkg/util/wait/wait.go:153\nk8s.io/apimachinery/pkg/util/wait.Until\n\tk8s.io/apimachinery@v0.17.2/pkg/util/wait/wait.go:88"}
```

`knative` returns a validation error, but user can not get this error, the isvc status will stay at `nill`
```
kubectl get inferenceservice 123a
NAME   URL   READY   DEFAULT TRAFFIC   CANARY TRAFFIC   AGE
123a                                                    7m13s
```

### PR DESCRIPTION
**Pull request by jazzsir**
. Check if ISVC name starts with alphabetical character.

Fixes #1059

**What this PR does / why we need it**:

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #

**Special notes for your reviewer**:

1. Please confirm that if this PR changes any image versions, then that's the sole change this PR makes.

**Release note**:

```release-note

```

### CODE REVIEW

File: pkg/apis/serving/v1alpha2/inferenceservice_validation.go [hunk 1]
```diff
@@ -32,6 +33,7 @@ const (
 	TrafficBoundsExceededError          = "TrafficPercent must be between [0, 100]."
 	TrafficProvidedWithoutCanaryError   = "Canary must be specified when CanaryTrafficPercent > 0."
 	UnsupportedStorageURIFormatError    = "storageUri, must be one of: [%s] or match https://{}.blob.core.windows.net/{}/{} or be an absolute or relative local path. StorageUri [%s] is not supported."
+    InvalidKSVCNameError                = "The InferenceService %s is invalid: the name must consist of lower case alphanumeric characters or '-', and must start with alphabetical character. (e.g. 'example-com')"
```
Comment (ellistarn) [inline pkg/apis/serving/v1alpha2/inferenceservice_validation.go:36]: Nit: your indentation is off. I recommend installing auto gofmt

**Review by ellistarn — COMMENTED** [review]
It's not clear to me why this name is necessary. Can you explain?

File: pkg/apis/serving/v1alpha2/inferenceservice_validation.go [hunk 2]
```diff
@@ -121,3 +128,10 @@ func validateCanaryTrafficPercent(spec InferenceServiceSpec) error {
 	}
 	return nil
 }
+
+func validateInferenceServiceName(isvc *InferenceService) error {
```
Comment (yuzisun) [inline pkg/apis/serving/v1alpha2/inferenceservice_validation.go:132]: suggest checking the logic here in kubernetes
https://github.com/kubernetes/kubernetes/blob/e077b0ffa48b1c0300a2cb25dd5d45d6fc358ade/staging/src/k8s.io/apimachinery/pkg/util/validation/validation.go#L187

File: pkg/apis/serving/v1alpha2/inferenceservice_validation.go (hunk 1 above)
Comment (jazzsir) [reply pkg/apis/serving/v1alpha2/inferenceservice_validation.go]: Thank you for the tip, I installed vim-go, cool!!

File: pkg/apis/serving/v1alpha2/inferenceservice_validation.go (hunk 2 above)
Comment (jazzsir) [reply pkg/apis/serving/v1alpha2/inferenceservice_validation.go:138]: Thanks, it has helped me a lot.

**Comment by jazzsir** [comment]
Moreover, isvc whose name includes '.' or exceeds 45 characters can't be deployed successfully. I applied these patterns.

File: pkg/apis/serving/v1alpha2/inferenceservice_validation.go [hunk 3]
```diff
@@ -121,3 +129,18 @@ func validateCanaryTrafficPercent(spec InferenceServiceSpec) error {
 	}
 	return nil
 }
+
+const isvcNameFmt string = "[a-z]([-a-z0-9]*[a-z0-9])?"
+const isvcNameMaxLength int = 45
+
+var isvcRegexp = regexp.MustCompile("^" + isvcNameFmt + "$")
```
Comment (yuzisun) [inline pkg/apis/serving/v1alpha2/inferenceservice_validation.go:136]: Can you help move these to the const and var section on the top?

File: pkg/apis/serving/v1alpha2/inferenceservice_validation.go (hunk 2 above)
Comment (yuzisun) [reply pkg/apis/serving/v1alpha2/inferenceservice_validation.go:138]: Cool! can you help add a few unit tests for the name check?

**Comment by yuzisun** [comment]
@jazzsir Thanks for the change! please also help port this change to v1beta1 API pkg

File: pkg/apis/serving/v1alpha2/inferenceservice_validation.go (hunk 2 above)
Comment (jazzsir) [reply pkg/apis/serving/v1alpha2/inferenceservice_validation.go:140]: Sure, I added them

File: pkg/apis/serving/v1alpha2/inferenceservice_validation.go (hunk 3 above)
Comment (jazzsir) [reply pkg/apis/serving/v1alpha2/inferenceservice_validation.go]: I did that, Is this find?

**Comment by jazzsir** [comment]
> @jazzsir Thanks for the change! please also help port this change to v1beta1 API pkg

Okey, I applied this changes to v1beta1
If you have further requests, please let me know.

**Comment by yuzisun** [comment]
@jazzsir Thanks for the contribution! Congratulation to your first PR!

/lgtm
/approve

**Pull request merged**